### PR TITLE
Periodic boundary conditions for auxiliary physics

### PR DESCRIPTION
**Pull request by oguevremont**
### Description

The Cahn-Hilliard physics solver did not allow periodic boundary conditions yet, as well as heat transfer and tracer physics.

### Solution

Using an approach similar to the VOF physics solver, Cahn-Hilliard, heat transfer and tracer physics now allow periodic boundary conditions. The periodic boundary conditions parameters for all auxiliary physics (including VOF) are now explicit; the syntax is the same as for the fluids dynamics solver.

### Testing

A new test has been added and all past tests still pass. 

- lethe-fluid/cahn_hilliard_periodic_boundaries
- lethe-fluid/heat_transfer_periodic_boundaries
- lethe-fluid/tracer_periodic_boundaries

### Documentation

The documentation has been adjusted to mention how to use periodic boundary conditions with auxiliary physics solvers.

### Checklist (will be removed when merged)
See [this page](https://chaos-polymtl.github.io/lethe/documentation/contributing.html#pull-requests) for more information about the pull request process.

Code related list:
- [x] All in-code documentation related to this PR is up to date (Doxygen format)
- [x] Copyright headers are present and up to date
- [x] Lethe documentation is up to date
- [x] Fix has unit test(s) (preferred) or application test(s), and restart files are in the generator folder
- [x] The branch is rebased onto master
- [x] Changelog (CHANGELOG.md) is up to date
- [x] Code is indented with indent-all and .prm files (examples and tests) with prm-indent

Pull request related list:
- [x] Labels are applied
- [x] There are at least 2 reviewers (or 1 if small feature) excluding the responsible for the merge
- [x] If this PR closes an issue or is related to a project, it is linked in the "Projects" or "Development" section
- [x] If the fix is temporary, an issue is opened
- [ ] The PR description is cleaned and ready for merge

### CODE REVIEW

File: applications_tests/lethe-fluid/cahn_hilliard_periodic_boundaries.output [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+   Volume of triangulation:      4
+   Number of Cahn-Hilliard degrees of freedom: 572
+   Number of active cells:       988
+   Number of degrees of freedom: 3168
+   Volume of triangulation:      4
+   Number of Cahn-Hilliard degrees of freedom: 2112
+
+*******************************************************************************
+Transient iteration: 1        Time: 0.0005   Time step: 0.0005   CFL: 0       
+*******************************************************************************
```
Comment (blaisb) [inline applications_tests/lethe-fluid/cahn_hilliard_periodic_boundaries.output:17]: What does this test verify? I understand you have an output and etc, but there's nothing that you are analyzing here... I'm not sure how you will be able to say that this test works or not.

File: applications_tests/lethe-fluid/cahn_hilliard_periodic_boundaries.prm [hunk 2]
```diff
@@ -0,0 +1,186 @@
+# SPDX-FileCopyrightText: Copyright (c) 2025 The Lethe Authors
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception OR LGPL-2.1-or-later
+
+# Listing of Parameters
+# ---------------------
+
+set dimension = 2
+
+# Simulation and IO Control
```
Comment (blaisb) [inline applications_tests/lethe-fluid/cahn_hilliard_periodic_boundaries.prm:9]: Can you add a tiny comment that explains what the test does :)

File: doc/source/parameters/cfd/boundary_conditions_multiphysics.rst [hunk 3]
```diff
@@ -153,27 +153,31 @@ For VOF boundary conditions (multiphase flow), the possible ``types`` are ``none
 Cahn-Hilliard
 ^^^^^^^^^^^^^^
 
-For Cahn-Hilliard boundary conditions, the available ``types`` are ``noflux`` (default), ``dirichlet``, ``angle_of_contact``, and ``free_angle``. The parameters for each type of Cahn-Hilliard boundary conditions are:
+For Cahn-Hilliard boundary conditions, the available ``types`` are ``none`` (default), ``noflux``, ``dirichlet``, ``angle_of_contact``, and ``free_angle``. The parameters for each type of Cahn-Hilliard boundary conditions are:
 
 .. code-block:: text
```
Comment (blaisb) [inline doc/source/parameters/cfd/boundary_conditions_multiphysics.rst:158]: Periodic should be here also no?

File: source/solvers/cahn_hilliard.cc [hunk 4]
```diff
@@ -1054,6 +1054,22 @@ CahnHilliard<dim>::setup_dofs()
...
+              id,
+              this->simulation_parameters.boundary_conditions
+                .periodic_neighbor_id.at(id),
+              this->simulation_parameters.boundary_conditions.periodic_direction
+                .at(id),
+              nonzero_constraints);
+          }
+      }
+
     for (auto const &[id, type] :
```
Comment (blaisb) [inline source/solvers/cahn_hilliard.cc:1073]: Would it not be just simply easier to have to manually specify the periodicity for every physics.
I feel it would be a lot more explicit if we had it this way instead of assuming.

**Review by blaisb — COMMENTED** [review]
Good work, but I think this is a good opportunity to change the way we process periodicity in the other physics. Why not declare it as a boundary condition like we do for Navier-Stokes. This way we can have hybrid Boundary conditions. Let's be as eexplicit as possible instead of implicit.

**Comment by oguevremont** [comment]
> Good work, but I think this is a good opportunity to change the way we process periodicity in the other physics. Why not declare it as a boundary condition like we do for Navier-Stokes. This way we can have hybrid Boundary conditions. Let's be as eexplicit as possible instead of implicit.

I agree, now is a good opportunity to do it. I'll change it!

**Comment by blaisb** [comment]
> > Good work, but I think this is a good opportunity to change the way we process periodicity in the other physics. Why not declare it as a boundary condition like we do for Navier-Stokes. This way we can have hybrid Boundary conditions. Let's be as eexplicit as possible instead of implicit.
> 
> I agree, now is a good opportunity to do it. I'll change it!

Thanks man, I really appreciate it

File: applications_tests/lethe-fluid/cahn_hilliard_periodic_boundaries.output (hunk 1 above)
Comment (oguevremont) [reply applications_tests/lethe-fluid/cahn_hilliard_periodic_boundaries.output]: I agree it was unclear. I made a different test, with a bubble that travels right and loops back with periodic boundary conditions. We follow the barycenter and see that it increases along X, then starts back at -1. If you like this I'll add something similar for tracer and heat transfer (VOF is already tested)

File: applications_tests/lethe-fluid/heat_transfer_periodic_boundaries.prm [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+# Initial conditions
+#---------------------------------------------------
+
+subsection initial conditions
+  subsection uvwp
+    set Function expression = 0;0;0
+  end
+  subsection temperature
+    set Function expression = if(x*x + y*y < 0.25,1,0)
+  end
```
Comment (blaisb) [inline applications_tests/lethe-fluid/heat_transfer_periodic_boundaries.prm:132]: So like advection of a blob of heat, legit, good test I like it

**Review by blaisb — APPROVED** [review]
I'll try 1-2 cases on my machine with it, but this seems very well made! Good job man!

File: doc/source/parameters/cfd/boundary_conditions_multiphysics.rst [hunk 6]
```diff
@@ -143,51 +164,60 @@ For VOF boundary conditions (multiphase flow), the possible ``types`` are ``none
...
 
-.. note::
-    For periodic boundary conditions, there is no need to specify anything in the ``boundary conditions VOF`` subsection. The periodic boundary condition must be specified in the ``boundary conditions`` subsection (see :doc:`boundary_conditions_cfd`).
-    
     
 Cahn-Hilliard
 ^^^^^^^^^^^^^^
 
-For Cahn-Hilliard boundary conditions, the available ``types`` are ``noflux`` (default), ``dirichlet``, ``angle_of_contact``, and ``free_angle``. The parameters for each type of Cahn-Hilliard boundary conditions are:
+For Cahn-Hilliard boundary conditions, the available ``types`` are ``none`` (default), ``noflux``, ``dirichlet``, ``angle_of_contact``, and ``free_angle``. The parameters for each type of Cahn-Hilliard boundary conditions are:
```
Comment (hepap) [inline doc/source/parameters/cfd/boundary_conditions_multiphysics.rst:175]:
```suggestion
For Cahn-Hilliard boundary conditions, the available ``types`` are ``none`` (default), ``noflux``, ``dirichlet``, ``angle_of_contact``, ``free_angle`` and ``periodic``. The parameters for each type of Cahn-Hilliard boundary conditions are:
```

File: examples/multiphysics/rayleigh-plateau-instability/rayleigh-plateau-J1-We020-Oh0_10_delta0_20/rayleigh-plateau-J1-We020-Oh0_10_delta0_20.prm [hunk 7]
```diff
@@ -149,14 +149,24 @@ end
...
+  subsection bc 1
+    set id                 = 2
+    set type               = periodic
+    set periodic_id        = 3
+    set periodic_direction = 1
+  end
+  subsection bc 2
+    set id   = 1
+    set type = none
+  end
```
Comment (hepap) [inline examples/multiphysics/rayleigh-plateau-instability/rayleigh-plateau-J1-We020-Oh0_10_delta0_20/rayleigh-plateau-J1-We020-Oh0_10_delta0_20.prm:169]: I'm not sure if this should be periodic, but I'm not sure which bc it should be... Maybe @AmishgaAlphonius would know?

File: source/solvers/vof_linear_subequations_solver.cc [hunk 8]
```diff
@@ -28,6 +28,24 @@ VOFLinearSubequationsSolver<dim>::setup_dofs()
...
+            this->dof_handler,
+            id,
+            this->simulation_parameters.boundary_conditions_vof
+              .periodic_neighbor_id.at(id),
+            this->simulation_parameters.boundary_conditions_vof
+              .periodic_direction.at(id),
+            this->constraints);
+        }
+    }
+
```
Comment (hepap) [inline source/solvers/vof_linear_subequations_solver.cc:48]: I'm not sure if periodic bc in VOF should be translated to periodic bc in the phase fraction gradient and curvature L2 projections. What do you think @blaisb @AmishgaAlphonius ?

**Review by hepap — CHANGES_REQUESTED** [review]
I left some things to check, but other than that, good work! I like the tests you added :)

File: source/solvers/vof_linear_subequations_solver.cc (hunk 8 above)
Comment (blaisb) [reply source/solvers/vof_linear_subequations_solver.cc:49]: I think they should :)

**Comment by blaisb** [comment]
@oguevremont Let me just merge the other CI PR, then I will rebase this PR and I will merge it afterwards :)

**Review by hepap — APPROVED** [review]
